examples/amm/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+use async_graphql::{Request, Response};
+use linera_sdk::{
+    base::{ContractAbi, ServiceAbi},
+    views::MapView,
+};
+use linera_views::views::ViewError;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub struct AMMAbi;
+
+impl ContractAbi for AMMAbi {
+    type InitializationArgument = ();
+    type Parameters = ();
+    type Operation = Operation;
+    type ApplicationCall = ();
+    type Message = ();
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for AMMAbi {
+    type Query = Request;
+    type QueryResponse = Response;
+    type Parameters = ();
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Operation {
+    Swap {
+        input_token: String,
+        output_token: String,
+        amount: u64,
+    },
+    AddLiquidity {
+        token: String,
+        amount: u64,
+    },
+    RemoveLiquidity {
+        token: String,
+        amount: u64,
+    },
+}
+
+pub fn calculate_output_amount(
+    input_amount: u64,
+    input_pool_balance: u64,
+    output_pool_balance: u64,
+) -> Result<u64, AMMError> {
+    if input_pool_balance == 0 || output_pool_balance == 0 {
+        return Err(AMMError::InvalidPoolBalanceError);
+    }
+
+    let input_pool_balance_float = input_pool_balance as f64;
+    let output_pool_balance_float = output_pool_balance as f64;
+
+    let input_amount_float = input_amount as f64;
+    let output_amount_float =
+        (input_amount_float * output_pool_balance_float) / input_pool_balance_float;
+
+    Ok(output_amount_float as u64)
+}
+
+pub async fn get_pool_balance(
+    token_pool: &MapView<String, u64>,
+    token: &String,
+) -> Result<u64, AMMError> {
+    if let Ok(option) = token_pool.get(token).await {
+        if let Some(balance) = option {
+            Ok(balance)
+        } else {
+            // get returned with no issues, but found nothing
+            Ok(0)
+        }
+    } else {
+        Err(AMMError::TokenNotFoundInPoolError)
```

Comment:
If the token isn't in the pool, isn't that the `Ok(None)` case?
If there's a view error, should we just return that, i.e. use `?` above?

Reply:
Yeah, that's what I was doing before, but in the test I check the balance to be 0 for the tokens before starting. Then this would fail because we haven't inserted liquidity for that token yet.
But I can also just assume in the test that the initial balance is zero and change this back